integrationTests/state/genesisState_test.go

```diff
@@ -0,0 +1,304 @@
+
+	_, _ = adbReference.RootHash()
+
+	noOfTests := 1000
+	for i := 0; i < noOfTests; i++ {
+		rootHash, adb := getRootHashByRunningInitialBalances(genesisBalances.InitialBalances)
+		if !bytes.Equal(rootHash, referenceRootHash) {
+			_, _ = adb.RootHash()
+			fmt.Printf("**** Wrong root hash on iteration %d: %s\n", i, base64.StdEncoding.EncodeToString(rootHash))
+			assert.Fail(t, "wrong root hash encountered")
+			return
+		}
+	}
+}
+
+func TestExtensionNodeToBranchVariant1(t *testing.T) {
+	marsh := &marshal.JsonMarshalizer{}
+	hasher := sha256.Sha256{}
+
+	tr1, _ := trie.NewTrie(createMemUnit(), marsh, hasher)
+	tr2, _ := trie.NewTrie(createMemUnit(), marsh, hasher)
+
+	key1 := "e353dd9e3da522eb366e751346195a10"
+	key2 := "a5dfc2ec3b0607e820ad375c5074c510"
+	key3 := "eb6d6e15652c0c4d1f73490e12c8b310"
+	val := "value"
+
+	_ = tr1.Update([]byte(key1), []byte(val))
+	_ = tr1.Update([]byte(key2), []byte(val))
+	_ = tr1.Update([]byte(key3), []byte(val))
+
+	fmt.Println()
+	tr1.Print(os.Stdout)
+	fmt.Println()
+
+	hash1, _ := tr1.Root()
+	fmt.Printf("root hash1: %s\n", base64.StdEncoding.EncodeToString(hash1))
+
+	_ = tr2.Update([]byte(key1), []byte(val))
+	_ = tr2.Update([]byte(key3), []byte(val))
+	_ = tr2.Update([]byte(key2), []byte(val))
+	hash2, _ := tr2.Root()
+	fmt.Printf("root hash2: %s\n", base64.StdEncoding.EncodeToString(hash2))
+
+	fmt.Println()
+	tr2.Print(os.Stdout)
+	fmt.Println()
+
+	assert.Equal(t, hash1, hash2)
+}
+
+func TestExtensionNodeToBranchVariant2(t *testing.T) {
+	marsh := &marshal.JsonMarshalizer{}
+	hasher := sha256.Sha256{}
+
+	tr1, _ := trie.NewTrie(createMemUnit(), marsh, hasher)
+	tr2, _ := trie.NewTrie(createMemUnit(), marsh, hasher)
+
+	key1 := "e353dd9e3da522eb366e751346195a10"
+	key2 := "eb6d6e15652c0c4d1f73490e12c8b310"
+	key3 := "6f1d4baa654871c773d0af8a4ccf0410"
+	key4 := "cdb0c9d63e94c56f18a75d1d186cac10"
+	key5 := "176aa1b6ff17951ba202c6e8bcdaf410"
+	key6 := "a5dfc2ec3b0607e820ad375c5074c510"
+	val := "value"
+
+	_ = tr1.Update([]byte(key5), []byte(val))
+	_ = tr1.Update([]byte(key3), []byte(val))
+	_ = tr1.Update([]byte(key1), []byte(val))
+	_ = tr1.Update([]byte(key6), []byte(val))
+	_ = tr1.Update([]byte(key2), []byte(val))
+	_ = tr1.Update([]byte(key4), []byte(val))
+
+	fmt.Println()
+	tr1.Print(os.Stdout)
+	fmt.Println()
+
+	hash1, _ := tr1.Root()
+	fmt.Printf("root hash1: %s\n", base64.StdEncoding.EncodeToString(hash1))
+
+	_ = tr2.Update([]byte(key1), []byte(val))
+	_ = tr2.Update([]byte(key2), []byte(val))
+	_ = tr2.Update([]byte(key3), []byte(val))
+	_ = tr2.Update([]byte(key4), []byte(val))
+	_ = tr2.Update([]byte(key5), []byte(val))
+	_ = tr2.Update([]byte(key6), []byte(val))
+
+	fmt.Println()
+	tr2.Print(os.Stdout)
+	fmt.Println()
+
+	hash2, _ := tr2.Root()
+	fmt.Printf("root hash2: %s\n", base64.StdEncoding.EncodeToString(hash2))
+
+	assert.Equal(t, hash1, hash2)
+}
+
+func TestExtensiveUpdatesAndRemovesWithConsistencyBetweenCylces(t *testing.T) {
```

Comment:
would move to a smokeTest folder. suggestion.

Reply:
TBD. Will move other tests (in another PR) as well if we decide to do that.